c2rust-analyze/src/rewrite/expr/unlower.rs

```diff
@@ -0,0 +1,379 @@
+            }
+
+            hir::ExprKind::Call(_, args) | hir::ExprKind::MethodCall(_, args, _) => {
+                let (loc, _mir_pl, _mir_func, mir_args) = match self.get_last_call(&locs) {
+                    Some((l, pl, f, a)) if self.is_var(pl) => (l, pl, f, a),
+                    _ => {
+                        warn(self, "expected final Call to store into var");
+                        return;
+                    }
+                };
+                self.record_desc(loc, &[], ex, MirOriginDesc::StoreIntoLocal);
+                self.record(loc, &[SubLoc::Rvalue], ex);
+                for (i, (arg, mir_arg)) in args.iter().zip(mir_args).enumerate() {
+                    self.record(loc, &[SubLoc::Rvalue, SubLoc::CallArg(i)], arg);
+                    // TODO: distribute extra `locs` among the various args
+                    self.visit_expr_operand(arg, mir_arg, &[]);
+                }
+                if locs.len() > 1 {
+                    warn!("NYI: extra locations {:?} in Call", &locs[..locs.len() - 1]);
+                }
+            }
+
+            _ => {
+                // For all other `ExprKind`s, we expect the last `loc` to be an assignment storing
+                // the final result into a temporary.
+                let (loc, _mir_pl, mir_rv) = match self.get_last_assign(&locs) {
+                    Some((l, pl, r)) if self.is_var(pl) => (l, pl, r),
+                    _ => {
+                        warn(self, "expected final Assign to store into var");
+                        return;
+                    }
+                };
+                self.record_desc(loc, &[], ex, MirOriginDesc::StoreIntoLocal);
+                self.record(loc, &[SubLoc::Rvalue], ex);
+                self.visit_expr_rvalue(ex, mir_rv, &locs[..locs.len() - 1]);
+            }
+        }
+    }
+
+    fn visit_expr_rvalue(
+        &mut self,
+        ex: &'tcx hir::Expr<'tcx>,
+        _rv: &'a mir::Rvalue<'tcx>,
+        _extra_locs: &[Location],
+    ) {
+        let _g = panic_detail::set_current_span(ex.span);
+        // TODO: handle adjustments, overloaded operators, etc
+    }
+
+    fn visit_expr_operand(
+        &mut self,
+        ex: &'tcx hir::Expr<'tcx>,
+        _rv: &'a mir::Operand<'tcx>,
+        _extra_locs: &[Location],
+    ) {
+        let _g = panic_detail::set_current_span(ex.span);
+        // TODO: handle adjustments, overloaded operators, etc
+    }
+}
+
+impl<'a, 'tcx> intravisit::Visitor<'tcx> for UnlowerVisitor<'a, 'tcx> {
+    type NestedFilter = nested_filter::OnlyBodies;
+
+    fn nested_visit_map(&mut self) -> Self::Map {
+        self.tcx.hir()
+    }
+
+    fn visit_expr(&mut self, ex: &'tcx hir::Expr<'tcx>) {
+        self.visit_expr_inner(ex);
+        intravisit::walk_expr(self, ex);
+    }
+}
+
+pub fn unlower<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    mir: &Body<'tcx>,
+    hir_body_id: hir::BodyId,
+) -> BTreeMap<(Location, Vec<SubLoc>), MirOrigin> {
```

Comment:
Is a `BTreeMap` map used here to sort by `(Location, Vec<SubLoc>)`?  If so, could you add that to the docs?

Reply:
Done (though currently we only rely on this feature for the debug prints in this function)

Reply:
That's a lot clearer, thanks!  Especially the part about prefixes and looking up by `Location` only for any `SubLoc`s.